File: Sources/YToolsCore/ApplicationAliasMatcher.swift
import Foundation

/// Parses and ranks user-defined application aliases without touching the
/// filesystem. The launcher can therefore keep alias matching in the same
/// fast, local path as ordinary application-name matching.
public struct ApplicationAliasMatcher: Sendable {
    private let normalizer = SearchTextNormalizer()

    public init() {}

    public func aliases(from rawValue: String) -> [String] {
        rawValue
            .components(separatedBy: CharacterSet(charactersIn: ",，;；\n"))
            .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
            .filter { !$0.isEmpty }
    }

    public func score(query: String, aliases: [String]) -> Int? {
        let queryForms = normalizer.forms(for: query)
        guard !queryForms.normalized.isEmpty else { return nil }
        return aliases.compactMap { score(query: queryForms, alias: $0) }.max()
    }

    private func score(query: SearchTextForms, alias: String) -> Int? {
        let candidate = normalizer.forms(for: alias)
        if candidate.normalized == query.normalized { return 960 }
        if candidate.transliteration == query.normalized { return 940 }
        if candidate.normalized.hasPrefix(query.normalized) { return 880 }
        if candidate.transliteration.hasPrefix(query.normalized) { return 860 }
        if candidate.abbreviation.hasPrefix(query.normalized)
            || candidate.transliterationInitials.hasPrefix(query.normalized) { return 830 }
        if candidate.normalized.contains(query.normalized)
            || candidate.transliteration.contains(query.normalized) { return 760 }
        guard query.normalized.count >= 2 else { return nil }
        let normalizedFuzzy = normalizer.fuzzyScore(
            query: query.normalized,
            candidate: candidate.normalized
        )
        let transliteratedFuzzy = normalizer.fuzzyScore(
            query: query.normalized,
            candidate: candidate.transliteration
        )
        return [normalizedFuzzy, transliteratedFuzzy]
            .compactMap { $0 }
            .max()
            .map { 620 + $0 }
    }
}

File: src/YTools.Windows/Core/ApplicationAliasMatcher.cs
namespace YTools.Core;

/// <summary>
/// Parses and ranks user-defined application aliases without touching the
/// filesystem, so alias matching stays in the same fast local path as names.
/// </summary>
public sealed class ApplicationAliasMatcher
{
    private static readonly char[] Separators = { ',', '，', ';', '；', '\n' };
    private readonly SearchTextNormalizer _normalizer = new();

    public IReadOnlyList<string> Aliases(string rawValue)
    {
        return rawValue
            .Split(Separators, StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries)
            .Where(alias => !string.IsNullOrEmpty(alias))
            .ToList();
    }

    public int? Score(string query, IReadOnlyList<string> aliases)
    {
        var queryForms = _normalizer.Forms(query);
        if (string.IsNullOrEmpty(queryForms.Normalized))
        {
            return null;
        }

        return aliases
            .Select(alias => Score(queryForms, alias))
            .Where(score => score is not null)
            .DefaultIfEmpty(null)
            .Max();
    }

    private int? Score(SearchTextForms query, string alias)
    {
        var candidate = _normalizer.Forms(alias);
        if (candidate.Normalized == query.Normalized)
        {
            return 960;
        }

        if (candidate.Transliteration == query.Normalized)
        {
            return 940;
        }

        if (candidate.Normalized.StartsWith(query.Normalized, StringComparison.Ordinal))
        {
            return 880;
        }

        if (candidate.Transliteration.StartsWith(query.Normalized, StringComparison.Ordinal))
        {
            return 860;
        }

        if (candidate.Abbreviation.StartsWith(query.Normalized, StringComparison.Ordinal)
            || candidate.TransliterationInitials.StartsWith(query.Normalized, StringComparison.Ordinal))
        {
            return 830;
        }

        if (candidate.Normalized.Contains(query.Normalized, StringComparison.Ordinal)
            || candidate.Transliteration.Contains(query.Normalized, StringComparison.Ordinal))
        {
            return 760;
        }

        if (query.Normalized.Length < 2)
        {
            return null;
        }

        var normalizedFuzzy = _normalizer.FuzzyScore(query.Normalized, candidate.Normalized);
        var transliteratedFuzzy = _normalizer.FuzzyScore(query.Normalized, candidate.Transliteration);
        var best = new[] { normalizedFuzzy, transliteratedFuzzy }.Where(score => score is not null).Max();
        return best is { } fuzzy ? 620 + fuzzy : null;
    }
}
